isaac/controllers.md

```diff
@@ -0,0 +1,255 @@
+If a resource is neither created nor consumed, it is _not yet created_.
+As part of their immutable data, resources can have complex proof obligations (which we call _resource logics_) determining when they can be created or consumed, and these may depend on the state of other resources.
+A resource logic can for example, specify exactly what programs can consume this resource: it would require a proof that the resources created are precisely the outputs of running a particular program with the consumed resources as inputs.
+Such a proof might be as simple as a full execution trace of the program, or as complex as a [zero-knowledge proof](https://eprint.iacr.org/2021/370).
+Through these logics, resources can encode fairly arbitrary state, not limited to scalar registers or tokens, while still allowing [ZKP-style confidential transactions](https://zenodo.org/doi/10.5281/zenodo.10498990).
+Crucially, resource logics can also constrain _where a resource can be transferred_: not all resources will be able to be transferred to all controllers. 
+
+### Controller Tags
+Each resource's _controller tag_ is part of this immutable data.
+It details which controllers define this resource's state.
+Resources transition between these states in transactions ordered by controllers.
+Each resource's tag therefore specifies a single controller that can order transactions for each type of transition, ensuring there is a single authority in charge of deciding whether each transition has or has not occurred:
+- The resource's _creating controller_ determines if the resource has been created.
+- The resource's _terminal controller_ determines if the resource has been consumed.
+Transactions which perform a state transition but are ordered by the wrong controller are _invalid_.
+
+
+### Transactions
+All state updates occur in [Transactions](https://en.wikipedia.org/wiki/Database_transaction): atomic updates that designate a set of _input_ resources (which must be previously _created_) that are _consumed_, and a set of _output_ resources (which must not yet have been created) which are _created_.
+Transactions can only update state controlled by one controller, and must include checkable proofs that the relevant resource logics of each resource created or consumed are satisfied.
+However, input resources may have been _created_ in a transaction on another controller.
+Therefore, controllers can sync with one another, allowing transactions to check if resources on other controllers have been created.
+These updates can be asynchronous, so it is possible a transaction will not immediately be able to prove that a resource has been created.
+
+
+### Controllers
+Controllers order transactions: each controller decides on an ever-growing sequence of transactions defining the execution _trace_, and thus the current state, of a state machine.
```

Comment:
going thought the report, thought this was the strongest one sentence definition of a controller. maybe worth elevating.